### PR TITLE
Read request packages and dependencies from packages.json

### PR DESCRIPTION
**Pull request by tkdchen**
CLBOUDBLD-4728
CLBOUDBLD-4729

Both endpoints /requests and /requests/id response the Request JSON
serialized from Request.to_json. Cachito starts to read a request's
packages and dependencies from the packages.json instead of from the
database.

Signed-off-by: Chenxiong Qi <cqi@redhat.com>

### CODE REVIEW

File: cachito/web/models.py [hunk 1]
```diff
@@ -890,31 +864,22 @@ def _state_to_json(state):
...
+            )
+            if bundle_dir.packages_data.exists():
+                packages_data = json.loads(bundle_dir.packages_data.read_bytes())
+                packages = packages_data.get("packages")
+                if packages:
+                    rv["packages"] = packages
+                    deps = []
+                    for package in packages_data["packages"]:
+                        deps.extend(package.get("dependencies", []))
+                    rv["dependencies"] = deps
```
Comment (chmeliik) [inline cachito/web/models.py:881]: The dependencies will need to be de-duplicated by `(name, version, type, dev)`

It would also be good to add unit tests where 2 different packages have the same dependency

Comment (brunoapimentel) [reply cachito/web/models.py:881]: I think it's also worth mentioning that the dependencies will need to be sorted here as well (once #481 is merged, that is).

File: cachito/web/utils.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 from collections import OrderedDict
+from collections.abc import Iterator
+from typing import Any, Dict, List
```
Comment (chmeliik) [inline cachito/web/utils.py:4]: If the `Iterator` import is only for typing, could it be imported from the `typing` modules instead?

File: cachito/web/models.py [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
+from cachito.web.utils import unique_packages
```
Comment (chmeliik) [inline cachito/web/models.py:2]: Could this be moved to the other `cachito` imports?

File: cachito/web/models.py [hunk 4]
```diff
@@ -893,35 +868,25 @@ def _state_to_json(state):
...
-                )
-                for req_package in self.request_packages
-            ]
-
-            sort_packages_and_deps_in_place(rv["packages"])
+            bundle_dir = RequestBundleDir(
+                self.id, root=flask.current_app.config["CACHITO_BUNDLES_DIR"]
+            )
+            if bundle_dir.packages_data.exists():
+                packages_data = json.loads(bundle_dir.packages_data.read_bytes())
```
Comment (chmeliik) [inline cachito/web/models.py:878]: Would it make sense to use the `PackagesData` class to load this data instead? The class could even get a new method for generating the top-level dependencies list

File: tests/test_api_v1.py [hunk 5]
```diff
@@ -2439,3 +2303,160 @@ def test_filter_requests(
...
+                    {"name": "async", "type": "npm", "version": "1.2.0"},
+                    {"name": "underscore", "type": "npm", "version": "1.13.0"},
+                ],
+            },
+        ],
+    }
+
+    expected_deps = [
+        {"name": "d1", "type": "go-package", "version": "1"},
+        {"name": "d2", "replaces": None, "type": "go-package", "version": "2"},
```
Comment (chmeliik) [inline tests/test_api_v1.py:2431]: Seeing this, I realized that Cachito currently always sets `"replaces": None` for dependencies that do not have replacements.

https://github.com/release-engineering/cachito/blob/ece43b86e31e7f858bd905cd58ed3887bd7e52d4/cachito/web/models.py#L477-L478
https://github.com/release-engineering/cachito/blob/ece43b86e31e7f858bd905cd58ed3887bd7e52d4/cachito/web/models.py#L903-L906

I'm not sure how much we need to worry about this detail of the API response, but if we want to be 100% backwards compatible, we need to do the same.

File: cachito/web/models.py [hunk 6]
```diff
@@ -893,35 +868,25 @@ def _state_to_json(state):
...
-                    subpath=req_package.subpath,
-                )
-                for req_package in self.request_packages
-            ]
-
-            sort_packages_and_deps_in_place(rv["packages"])
+            bundle_dir = RequestBundleDir(
+                self.id, root=flask.current_app.config["CACHITO_BUNDLES_DIR"]
+            )
+            if bundle_dir.packages_data.exists():
```
Comment (brunoapimentel) [inline cachito/web/models.py:877]: I will need to load the packages data as well to generate the content manifest (CLOUDBLD-4730), so I was thinking that it could be a good idea to make this reusable. I have two suggestions on how it could be made:

1- Load the package data from the json file right after the object is loaded from the database. This way the data would always be available and its behavior would be similar to having the packages on the database.

2- Have this behavior be done on private function on the class, so it would be easy to just reuse this part of the code. This way, it's easier to implement and we'd have to worry less about corner cases of suggestion 1.

Comment (tkdchen) [reply cachito/web/models.py:877]: I have changes to the `PackagesData` locally, I'll push it later and please review it in advance.

File: tests/test_api_v1.py (hunk 5 above)
Comment (tkdchen) [reply tests/test_api_v1.py:2424]: @chmeliik 

Is this already covered in every `fetch_*_source` when writing the packages data? The resolved dependencies should have replaces, and they will be written into the package manager specific packages data file and then gathered into the `packages.json`.

Comment (chmeliik) [reply tests/test_api_v1.py:2424]: I don't think so. The dependencies returned by the `resolve_*` functions may or may not have the "replaces" key. In fact, I think only deps from `resolve_gomod` have the key.

File: cachito/utils.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                return
+            for p in packages:
+                self.add_package(p, p.get("path", os.curdir), p["dependencies"])
+
+    def sort(self) -> None:
+        """Sort packages and each package's dependencies in place."""
+        sort_packages_and_deps_in_place(self._packages)
+
+    def all_dependencies(self) -> List[Dict[str, Any]]:
+        """Gather dependencies together from very package.
```
Comment (chmeliik) [inline cachito/utils.py:161]:
```suggestion
        """Gather dependencies together from every package.
```

File: cachito/utils.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+        :return: a list of sorted and deduplicated dependencies gathered from
+            every package. If no package is added, an empty list will be returned.
+        :rtype: list[dict[str, any]]
+        """
+        return list(
+            unique_packages(
+                sorted(
+                    itertools.chain(*[pkg["dependencies"] for pkg in self._packages]),
```
Comment (chmeliik) [inline cachito/utils.py:170]:
```suggestion
                    itertools.chain.from_iterable(pkg["dependencies"] for pkg in self._packages),
```

File: cachito/workers/tasks/utils.py [hunk 9]
```diff
@@ -20,9 +18,6 @@
     "get_request",
     "get_request_state",
     "set_request_state",
-    "set_packages_and_deps_counts",
```
Comment (chmeliik) [inline cachito/workers/tasks/utils.py:23]: I think this should stay here

File: tests/test_utils.py [hunk 10]
```diff
@@ -72,3 +76,264 @@ def test_deep_sort_icm(orig_items):
...
+        {"name": "pkg6", "type": "pip", "version": "1.0.0", "dev": False},
+        {"name": "pkg5", "type": "pip", "version": "1.0.0", "dev": True},
+    ]
+
+    sort_packages_and_deps_in_place(packages)
+
+    assert packages == sorted_packages
+
+
+class TestPackagesData:
```
Comment (chmeliik) [inline tests/test_utils.py:176]: Could we have a test for the all_dependencies method?

File: cachito/web/models.py [hunk 11]
```diff
@@ -893,35 +866,14 @@ def _state_to_json(state):
...
-                for req_package in self.request_packages
-            ]
-
-            sort_packages_and_deps_in_place(rv["packages"])
+            bundle_dir = RequestBundleDir(
+                self.id, root=flask.current_app.config["CACHITO_BUNDLES_DIR"]
+            )
+            packages_data = PackagesData()
+            packages_data.load(bundle_dir.packages_data)
+            packages_data.sort()
```
Comment (chmeliik) [inline cachito/web/models.py:874]: Technically we don't need to sort here anymore, when loading a single sorted file, the packages data will be sorted.

We included the sorting changes in the same release as the packages.json generation, so there should not be any unsorted files in any of the Cachito environments.

**Review by chmeliik — COMMENTED** [review]
+1 from moving the `PackagesData` class to a shared module, I just have a suggestion about the naming of that module. I would also like that change to have its own commit, but it might be difficult to split them up now so it's not a big deal.

File: cachito/web/models.py (hunk 6 above)
Comment (brunoapimentel) [reply cachito/web/models.py]: I like it, I think it looks a lot better now :)

File: cachito/utils.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        contain four keys, type, dev, name and version.
+    :type package: dict[str, any]
+    :return: a four-elements tuple as the sort key in the order of type, dev,
+        name and version.
+    :rtype: tuple[str, bool, str, str]
+    """
+    return package["type"], package.get("dev", False), package["name"], package["version"]
+
+
+def sort_packages_and_deps_in_place(packages: List[Dict[str, Any]]) -> None:
```
Comment (brunoapimentel) [inline cachito/utils.py:28]: Besides some unit tests, it seems this function is now only being used inside the PackageData class.

If we could work around the tests, it would be possible to make it a private function inside the PackageData class, and this way we'd have less `utils` going around.

We don't need to do this right now, though, it's just a thought.

File: cachito/web/models.py [hunk 13]
```diff
@@ -16,10 +16,11 @@
 import urllib.parse
 from werkzeug.exceptions import Forbidden
 
-from cachito.web import content_manifest
 from cachito.errors import ContentManifestError, ValidationError
+from cachito.paths import RequestBundleDir
+from cachito.web import content_manifest
 from cachito.web import db
-from cachito.workers.tasks.utils import sort_packages_and_deps_in_place
+from cachito.utils import PackagesData
```
[the remark this answers is not in the document]
Comment (tkdchen) [reply cachito/web/models.py:23]: Looks errors.py and paths.py can also be moved into the common module. I'll make this change in a separate commit.

**Review by chmeliik — COMMENTED** [review]
I found a regression after running integration tests. All requests now have `packages_count = 0, dependencies_count = 0` set in the database. I think the reason is that in the `finalize_request()` task, `get_request()` is called before `aggregate_packages_data()`. That used to work before, but not anymore

**Review by chmeliik — COMMENTED** [review]
LGTM :+1:

Are you planning to rename the `cachito.utils` module in this PR or a different one?

**Comment by tkdchen** [comment]
> Are you planning to rename the cachito.utils module in this PR or a different one?

@chmeliik Yes. The changes are in local branch already. When this PR is good enough and merged, I'll rebase the local changes and push for review.

**Review by chmeliik — APPROVED** [review]
